providers/mlx5/mlx5.c

```diff
@@ -388,44 +377,46 @@ int mlx5_destroy_psv(struct mlx5_psv *psv)
 	return ret;
 }
 
+#if defined(__i386__)
+static unsigned int get_mask(int startbit, int endbit)
+{
+	int i;
+	int result = 0;
+
+	for (i = startbit; i >= endbit; i--)
+		result |= (1 << i);
+
+	return result;
+}
+
+#define EXTRACT_BITS(val, start, end) (((val) & get_mask((start), (end))) >> (end))
+#endif
+
 static int mlx5_is_sandy_bridge(void)
 {
-	char line[128];
-	FILE *fd;
-	int rc = 0;
-	int cur_cpu_family = -1;
-	int cur_cpu_model = -1;
+#if defined(__i386__)
+	unsigned int ax, bx, cx, dx;
+	unsigned int ext_family_id, ext_model_id, family_id, model_id;
 
-	fd = fopen("/proc/cpuinfo", "r");
-	if (!fd)
+	if (!__get_cpuid(1, &ax, &bx, &cx, &dx))
 		return 0;
 
-	while (fgets(line, 128, fd)) {
```

Comment:
This code is called once upon driver startup and is here for years, what is the motivation just now to optimize it ?